tsan fix browsethread

m0dB opened the pull request:
Fixes #13861 and improves the code quality of BrowseThread a bit.

Commit:
fix tsan data race with BrowseThread and improve the code quality

m0dB commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
     start(QThread::LowPriority);
-
+    qDebug() << "Wait to start browser background thread";
+    // Wait until the thread is running
```
this fixes the data race

m0dB replied:
Confirmed. This is the fix for the data race. Connecting the signals/slots while the thread is not yet running is the issue.

m0dB commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
-    m_path_mutex.lock();
+void BrowseThread::requestPopulateModel(mixxx::FileAccess path, BrowseTableModel* client) {
+    // Inform the thread to populate a new path.
+    // Note: if another request is currently being processed, this will replace it.
```
and that seems weird to me. it makes sense when the client is the same, but what if it's a different client? anyway, I didn't want to change the behaviour.

m0dB replied:
That makes sense.

m0dB commented on src/library/browse/browsethread.cpp:
```diff
@@ -144,15 +171,19 @@ void BrowseThread::populateModel() {
-
-        if (thisPath.info() != newPath.info()) {
-            qDebug() << "Abort populateModel()";
-            populateModel();
```
I didn't like this recursion, better give the control back to the run() method to handle the new request.

Commit:
typo

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
-    m_model_observer = nullptr;
-    //start Thread
+BrowseThread::BrowseThread(QObject* parent)
+        : QThread{parent}, m_modelObserver{}, m_bRun{} {
```
Not your fault: We use a p prefix for pointers and normally stack the initializer list. Would be nice if you adjust this as well.

m0dB replied:
will do

daschuer reviewed (commented):
Some comments. I have not yet fully understand the issue.

m0dB commented:
> Some comments. I have not yet fully understand the issue.

To explain the issue: tsan reported a data race in browsetablemodel.cpp:163 when connecting the table model with the result from BrowseThread::getInstanceRef(). IIRC the fix is waiting until the thread is started before returning from getInstanceRef (in the BrowseThread constructor) but I also changed getInstanceRef itself because the placing of the mutex didn’t seem correct to me. 

The other changes are mainly esthetics / code clarity.

I do have my doubts about this code, it seems strange to me that a new populate request for any browsetablemodel aborts the running request. But I didn’t want to change to behavior.

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
+    m_mutex.lock();
+    while (!m_bRun) {
+        m_condition.wait(&m_mutex);
+    }
```
I have not understood the loop. Isn't there only one event/codition we are waiting for? m_bRun shall become a semaphore? On the other hand syncronizing the thread start by halting the main thread feels wrong. It would be better to protect the shared sources explicit. What are the shared resources in this case we have tsan complains?

m0dB replied:
Yes, we are waiting for m_bRun to be set to true, as set by the thread::run method. A semaphore would work as well, but since we already have the condition, I take advantage of it.

daschuer replied:
> A semaphore would work as well, but since we already have the condition, I take advantage of it.

This reusing makees the code harder to read. So I would prefer two primitived for two tasks. 

The ThreadSanitizer warning is really confusing, what does it check for? Whether a variable is written/read from two threads? It must be a control structure in Qt. However waiting for the thread to start should not change the issue for my undertsanding. 

But if it is the solution i don't mind.

daschuer replied:
How does the solution work? I am afraid instead of a solution, we trick ThreadSanitizer.
Can, it happen that it treated the unrelated mutex as a proper guard?
Qt allows to send signals and reconnect them concurrently. So it should not be a problem. 
Maybe only a missing atomic annotation in Qt? 

Is there a way to find out what 0x0001466d5f28 is and which mutexted I are considered? 

Or is it s QT bug? Is 2.4 with Qt 5 affected as well.

daschuer replied:
Got it, I n this thread they discussed exactly our problem. Solution is to build Qt wir sanitizer flag: 
https://lists.qt-project.org/pipermail/interest/2022-November/038561.html

Conclusion here: We can safely ignore the issue.

m0dB replied:
I have build Qt with the sanitizer flag, though I used the normal way (passing the -fsanitize=thread flag), so unless this configure option does something else (I doubt it), the problem is still there.

I don't see you problem with using one and the same mutex and condition.

m0dB replied:
Anyway, I will see if I can build qt in my vcpkg with this configure option, and if that makes a difference. But I would appreciate it if you can point me to where I need to change this, given my inexperience with vcpkg.

daschuer replied:
I think you can add it here: 
https://github.com/mixxxdj/vcpkg/blob/74ff6c7ceb657e9f099626b5834884184209bf50/ports/qt5-base/portfile.cmake#L94

m0dB replied:
Ok, I will give that a try.

m0dB replied:
I am building against Qt6 though

m0dB replied:
which seems to use a different approach... Any idea how to do it there?

m0dB replied:
Anyway, I am now building Qt5. But it is on purpose that the 2.5 branch of mixxx vcpkg has Qt6 as the default? I am confused now as to what we are supposed to use for Mixxx 2.5.

m0dB replied:
I am running into lots of problems trying to build against Qt5, so it would be much appreciated if you could help me building Qt6 with vcpkg with this Qt sanitize thread option.

daschuer replied:
Mixxx 2.4.2 will be our last Qt5 release. So there is IMHO no need to deal with QT5 in this case. We just keep the Qt5 support in our 2.5 branch just in case we want to test regresions. 

I suggest to issue a PR against our vcpkg branch and use the build package downloaded from GitHub. For my undersanding you jsut need to add the sanitizer flag at the position I pointed out above.

m0dB replied:
Yes, I prefer to use Qt 6, but you pointed it out in the qt5 portfile.cmake. The qt6 portfile.cmake is completely different, and I don't see where to add this option.

m0dB replied:
It looks like this works. I'd prefer to set this from my triplets file arm64-osx-tsan.cmake, but for now this have to do.

diff --git a/ports/qtbase/portfile.cmake b/ports/qtbase/portfile.cmake
index 8b9e3b0d3..4a96412d7 100644
--- a/ports/qtbase/portfile.cmake
+++ b/ports/qtbase/portfile.cmake
@@ -92,7 +92,7 @@ INVERTED_FEATURES
     )
 
 list(APPEND FEATURE_OPTIONS -DCMAKE_DISABLE_FIND_PACKAGE_Libudev:BOOL=ON)
-list(APPEND FEATURE_OPTIONS -DFEATURE_xml:BOOL=ON)
+list(APPEND FEATURE_OPTIONS -DFEATURE_xml:BOOL=ON -DFEATURE_sanitize_thread:BOOL=ON)

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
+            m_condition.wait(&m_mutex);
+        }
+        auto path = std::move(m_path);
+        auto modelObserver = m_modelObserver;
```
Do we have a gurantee that the m_modelObserver outlives the thread?

m0dB replied:
yes, because it's holds a shared pointer to it.

daschuer replied:
m_modelObserver holds a share pointer to the Browsthread? Maybe you can add a comment to the code?

m0dB replied:
There is a single, refcounted, BrowseThread. This is what is returned by BrowseThread::getInstanceRef();

In BrowseTableModel.cpp it is stored: m_pBrowseThread = BrowseThread::getInstanceRef();

So, when all BrowseTableModel instances are gone, the ref count is 0, and the BrowseThread instance is destroyed.

But I am not sure why I have to explain this or add more comments. This is not my code! I only want to solve the thread sanitizer warnings. If you feel the (existing!) code needs more explanation, go ahead.

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -112,29 +141,27 @@ class YearItem: public QStandardItem {
-    auto thisPath = m_path;
-    BrowseTableModel* thisModelObserver = m_model_observer;
-    m_path_mutex.unlock();
+void BrowseThread::populateModel(mixxx::FileAccess path, BrowseTableModel* modelObserver) {
```
p prefix would be nice: pModelObserver

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
+    // Inform the constructor the thread started
     m_mutex.lock();
+    m_bRun = true;
+    m_condition.wakeOne();
```
wakeOne() and wait() below is confusing, Let's use a second sync method.

m0dB replied:
I put the request in a separate method.

daschuer commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
-    m_path_mutex.unlock();
-    m_locationUpdated.wakeAll();
+    m_modelObserver = modelObserver;
+    m_condition.wakeOne();
```
What can happen if the thread is not yet waiting? In this case no one is waked up. 
But the `m_modelObserver` and `m_path` have a value. I think the thread can first check if m_modelObserver is not null and pick that up before waiting. Probably the waiting in the constructor is obsolete than?

m0dB replied:
if m_path has a value, the thread will not enter the wait state.

m0dB commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
+    static QWeakPointer<BrowseThread> s_weakInstanceRef;
+    static QMutex s_mutex;
+
+    s_mutex.lock();
```
Ok, I will check again where the issue comes from, here or the thread being used before being started.

m0dB replied:
You are right. The original code was correct. I have reverted it, except for making 
```
   static QWeakPointer<BrowseThread> s_weakInstanceRef;
   static QMutex s_mutex;
```
local variables in the function instead of globals.

Please resolve if you are okay with the latest version.

m0dB commented on src/library/browse/browsethread.cpp:
```diff
@@ -14,80 +14,109 @@ namespace {
 
-void BrowseThread::executePopulation(mixxx::FileAccess path, BrowseTableModel* client) {
-    m_path_mutex.lock();
+void BrowseThread::requestPopulateModel(mixxx::FileAccess path, BrowseTableModel* modelObserver) {
```
will do

m0dB commented on src/library/browse/browsethread.h:
```diff
@@ -41,16 +41,13 @@ class BrowseThread : public QThread {
     QMutex m_mutex;
-    QWaitCondition m_locationUpdated;
-    volatile bool m_bStopThread;
+    QWaitCondition m_condition;
```
the thing is, I am using it for 2 things: waiting for the thread to be started and waiting for a new populate request.

m0dB replied:
I have renamed the member vars for clarity. Please review and resolve.

Commit:
act on review

Commit:
fix order

Commit:
stack initializers

Commit:
fix clazy

m0dB commented:
I have checked with Qt 6 with thread sanitizer activated through the Qt feature flags, and it still has to same problem, so I would like to ask to merge this PR. I feel I addressed all comments.

Commit:
Merge branch '2.5' into tsan-fix-browsethread

Commit:
renaming functions for clarity

m0dB commented:
Is there is a reason not to merge this? I confirm that, including with a thread sanitizer enabled Qt, I get TSAN warnings unless I wait for the thread to have started.

daschuer commented:
I did a bit of a research and discovered a related QT issues. 

https://bugreports.qt.io/browse/QTBUG-100336
.. looks quite similar. My guess is that the different backtrace happens due to not optimized code. 

Can you please use the attached the testing code and confirm that this is actually our issue? 

This bug is fixed in: Qt v6.8.0 and planned to 6.5 the stable branch we use for Mixxx 2.5 
https://github.com/qt/qtbase/commit/75d82afa0d3aad9b4f9857e439535fc49c4616bc

I was able to trace it down to https://github.com/qt/qtbase/commit/34fe9232dbf6a9fe58ebc4c7680bb67d2f642c40
> receiver->d_func()->connections.loadRelaxed()->currentSender = previous
Where it has been portet from QAtomic in v5.14.0
But the issue exists way longer with the old API. 

The issue seems affect a bunch of atomic variables and whenever a connection is astabisched between threads. Most likely this fix fixes only one symptome of the issue and the issue will happen again with the other signals. 
Luckily this is currently "only" a linter issue and we do not have a real user issue with that, I would prefer to revert the quite bulky and run-time consuming solution here and just wait for the upstream fix. The other refactorings in this thread are welcome. 

An alternative solution for all signals would be to connect the the connections via "m_model_observer" from the worker thread itself, this way the affected variables are only touched by one thread. I consider this a nice refactoring that can remain even if the upstrem issue is fixed. But I don't want to put that work on you. You may also directky "QMetaObject::invokeMethod" without connecting.
